### ISSUE TITLE
Abstract and move ExternData/Data/Dim

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1165]
Basically the whole file `returnn/tf/util/data.py` and then also the class `ExternData` should be moved to a framework common place, maybe like `returnn/data`, maybe separare submodules there like `extern_data.py`, `data.py`, `dim.py`, `batch_info.py`.

Then all TF specific code should be removed. This should not be much. This is `auto_create_placeholders` and a couple of the `copy_...` functions.

`Data` and co was always designed in a way to be generic and not TF specific. Now that we work on PyTorch (#1120), it becomes relevant for that as well (https://github.com/rwth-i6/returnn/issues/1120#issuecomment-1286887444).

`Data` and `Dim` should ideally also be simplified, or most importantly the core object and the `__init__`. The `__init__` should execute very fast, as this would be required for eager mode. Also see #975.

**Comment by albertz** [comment]
> `Data` and `Dim` should ideally also be simplified, or most importantly the core object and the `__init__`. The `__init__` should execute very fast, as this would be required for eager mode. Also see #975.

I was thinking about this.

One obvious thing to do is to use `__slots__`. This will give some speedup. However, this is not really enough.

To get it fast, we should remove most of the attribs, and really keep only what is relevant. However, it will be difficult to remove a lot of the attribs while keeping it compatible to existing logic, which we need to keep to not break older setups. But having lots of attribs already makes it slow, because we need to have a complex `__init__`, assigning all the attribs, going through all the arguments. And it also makes all other functions slow, which need to check for those cases.

But let's assume we can break old setups, and would design this completely new from scratch. How would it look then? What do we really need to keep then?

Now if we made `__init__` fast and short for this case, can we still support all the old code? I think yes, by introducing just one single extra attrib, named `_extra` or sth like that, which will normally be `None`, and only if some of those old features are used, then it is set to some other structure which has all the old attribs. In the code, we need to check this `_extra` attrib, but this is just a single check, and then it can go for the fast path.

**Comment by albertz** [comment]
So, let's speak about the minimal requirements, breaking old setups, and then adding backwards compatibility via such `_extra`attrib.

About `Dim`:

* We really should clarify if we need the `same_as`/`_same_for_batch_ctx` logic, and also the `declare_same_as` (#975). I.e. check for all current usages and see if that is really needed.
* `kind`: Not sure if we really need this. I don't think any code should really depend on this.
* `description`: Yes, but maybe renamed to `name`. Also, behavior should be better defined. Does this needs to be unique? Is `None` allowed (as it is currently)?
* `dimension`: Yes but we should actually separate it: The actual dimension (shape[axis]), which might be static even with `dyn_size_ext`, and then the `size` if static, or otherwise `dyn_size_ext`.
* `dyn_size_ext`: Yes.
* `match_priority`: Probably yes. There is the rare (but important) edge case for squared matrices.
* `_creation_idx`: Maybe.
* `_dyn_size_same`: No.
* `derived_from_tag`: No.
* `derived_from_op`: No.
* `batch`/`control_flow_ctx`: No, because the same info is also in `dyn_size_ext`.
* `src_data`/`src_axis`: No.
* `_undefined`/`auto_generated`: No.
* `generic`/`special`: No.

**Comment by albertz** [comment]
Also, we should think about whether we need `BatchInfo`. Actually I don't like that this is special about the batch dim. The only real needed functionality here is the packing of sequences (`pack_sequence` / flattening) but this should be maybe more generic for `Dim`s, and otherwise I think we don't really need it. Currently it is also for `SearchBeam` but I think this should rather be also a separate `Dim`. But when we do this, we still need to keep some of the extra info of `SearchBeam` like `dependency`, and maybe also unify it with `SearchChoices`, which has more, such as `beam_scores`, `owner`, `src_beams`.

**Comment by albertz** [comment]
About `Data`:

* `name`: Yes.
* `_placeholder`: Yes. Maybe rename to `data`, `tensor`, `native_tensor` or so.
* `dtype`: Yes. Maybe the type should be changed?
* `_dim_tags`: Yes.
* `sparse_dim`: Yes.
* `control_flow_ctx`: Maybe.
* `available_for_inference`: Unclear.
* `_batch`: No.
* `_beam`: No. Obsolete via `_batch` or dim tags and extra dim for beam.
* `_feature_dim_axis`: No.
* `time_dim_axis`: No.

**Comment by albertz** [comment]
When optimizing for performance, we should prioritize eager-mode frameworks (e.g. PyTorch), as the performance is most critical for those. This is an important point, because in those frameworks, we don't need control-flow-context logic.

**Comment by albertz** [comment]
On code organization:

* There should be `__slots__` for the main important attribs, and then one `_extra`. And add `@property` wrappers to all the extra attribs.
* `__init__` should be stripped down to the main attribs, and get `kwargs`, and only if `kwargs` has some data, it needs further processing.
* There can be a mixin class for most of the extra stuff which is maybe distracting in the main class. To work well with type checkers (including PyCharm), we can do it like described [here](https://stackoverflow.com/questions/55209849/python-mixin-unresolved-attribute-reference-pycharm), with type annotations in the mixin class.

**Comment by patrick-wilken** [comment]
Maybe I should know, but what actually is a "sparse_dim"? As far as I know "sparse" in our code base refers to integer data that represents indices in some vocabulary (class indices). So a `Data` object / tensor as a whole is either sparse or not. How is "sparse" an attribute of a `Dim`, i.e. an axis?
And very related: why is the vocabulary an attribute of an axis? If the elements of a tensor represent class indices then it can have an associated vocabulary, but why assign it to one axis?

**Comment by albertz** [comment]
You can always think of such sparse data as representing a one-hot encoding of a dense vector. This is the reason why we call it "sparse" in the first place. The feature dimension of this dense vector, this is what is represented by the sparse_dim. It makes sense to represent it as a `Dim`, just like all the other dims, because even when we never explicitly create the one-hot dense vector, we still have this dim explicitly in other places, such as in the embedding matrix, or the softmax layer (if these are the targets). And a static (finite) `Dim` can also always represent a vocab. A vocab can always have a corresponding `Dim`. Note that the axis never really matters because we have those `Dim` objects. It does not matter whether your BPE vocab dim is in axis 0, 1 or 2, or sparse. It's always the same dim.

**Comment by patrick-wilken** [comment]
Thanks, makes sense. Saw you added a corresponding doc string. 👍
